sdks/python/apache_beam/ml/inference/pytorch_inference.py

```diff
@@ -100,6 +112,21 @@ def _convert_to_result(
   return [PredictionResult(x, y) for x, y in zip(batch, predictions)]
 
 
+def default_tensor_inference_fn(
+    batch: Sequence[torch.Tensor],
+    model: torch.nn.Module,
+    device: str,
+    inference_args: Optional[Dict[str,
+                                  Any]] = None) -> Iterable[PredictionResult]:
+  # torch.no_grad() mitigates GPU memory issues
+  # https://github.com/apache/beam/issues/22811
+  with torch.no_grad():
+    batched_tensors = torch.stack(batch)
+    batched_tensors = _convert_to_device(batched_tensors, device)
+    predictions = model(batched_tensors, **inference_args)
+    return _convert_to_result(batch, predictions)
```

Comment:
Looking at this, my only concern is that doing the simple thing (calling `model.generate(...)` instead of `model(...)` or something similar) is now harder to do performantly.

I wonder if there's a way to still make that easy - one option would be to add something like a convenience function `make_tensor_inference_override_fn(model_function: string)` that generates a lambda. So the user code would then look like: `my_tensor_inference_fn = make_tensor_inference_overide_fn('generate')`, which would set `my_tensor_inference_fn` to an anonymous function that does:

```
  with torch.no_grad():
    batched_tensors = torch.stack(batch)
    batched_tensors = _convert_to_device(batched_tensors, device)
    predictions = model.generate(batched_tensors, **inference_args)
    return _convert_to_result(batch, predictions)
```

Does that make sense/sound reasonable?

Reply:
The convenience function kind of works, although you get a funky named function defined within the scope of the convenience function. No multi-line lambdas will do that. The function winds up looking something like this:

```
with torch.no_grad():
      batched_tensors = torch.stack(batch)
      batched_tensors = _convert_to_device(batched_tensors, device)
      pred_fn = model.get_attr(model_fn)
      predictions = pred_fn(batched_tensors, **inference_args)
      return _convert_to_result(batch, predictions)
```

Although the more I think about it, the more I think we should just provide a `generate` function users can pass since we know using that routing instead is a major motivating example for this feature. There's also something to be sad for making `_convert_to_device` and `_convert_to_result` available to users as building blocks for their custom functions.

Reply:
> although you get a funky named function defined within the scope of the convenience function. No multi-line lambdas will do that. The function winds up looking something like this:

Yeah, that's what I was imagining here. Note that while this is funky for us, its a smooth user experience because they don't have to be aware of the get_attr call.

> Although the more I think about it, the more I think we should just provide a generate function users can pass since we know using that routing instead is a major motivating example for this feature.

I agree, though I think `generate` is just an example and we want this to be as easy as possible for similar functions as well. I'd probably argue that a generic `make_tensor_inference_override_fn` (maybe with better naming 😅) does that better than us trying to hit all the functions a user could provide. We don't necessarily have to do that, but if you disagree let's get more peoples' voices/opinions involved.

Reply:
We should make sure whatever we do, we shouldn't instantiate the model before the pipeline startup. If the model is large, instantiating model's object takes a lot of time and sometimes pipeline won't start because we have hit a limit (https://cloud.google.com/dataflow/quotas#:~:text=The%20default%20disk%20size%20is,for%20Dataflow%20Shuffle%20batch%20pipelines._

Reply:
The helper route isn't that bad, a little verbose on our end but it should make user routing easier. First run at it was just pushed.